k-distribution/k-tutorial/1_basic/21_symbolic_execution/README.md

```diff
@@ -1 +1,174 @@
+the K rewrite engine is concrete; in other words, contains no logical
+variables. The LLVM Backend is a **concrete execution** engine, meaning that
+it is only capable of rewriting concrete configurations.
+
+By contrast, the Haskell Backend performs **symbolic execution**, which is
+capable of rewriting any configuration, including those where parts of the
+configuration are **symbolic**, ie, contain variables or uninterpreted
+functions.
+
+## Unification
+
+Previusly, we have introduced the concept that K rewrite rules operate by means
+of pattern matching: the current configuration being rewritten is pattern
+matched against the left-hand side of the rewrite rule, and the substitution
+is used in order to construct a new term from the right-hand side. In symbolic
+execution, we use
+[unification](https://en.wikipedia.org/wiki/Unification_%28computer_science%29)
+instead of pattern matching. To summarize, unification behaves akin to a
+two-way pattern matching where both the configuration and the left-hand side
+of the rule can contain variables, and the algorithm generates a
+**most general unifier** containing substitutions for the variables in both
+which will make both terms equal.
+
+## Feasibility
+
+Unification by itself cannot completely solve the problem of symbolic
+execution. One task symbolic execution must perform is to identify whether
+a particular symbolic term is **feasible**, that is to say, that there actually
+exists a concrete instantiation of that term such that all the logical
+constraints on that term can actually be satisfied. The Haskell Backend
+delegates this task to [Z3](https://github.com/Z3Prover/z3), a 
+[SMT solver](https://en.wikipedia.org/wiki/Satisfiability_modulo_theories).
+This solver is used to periodically trim configurations that are determined
+to be mathematically infeasible.
+
+## Symbolic terms
+
+The final component of symbolic execution consists of the task of introducing
+symbolic terms into the configuration. This can be done one of two different
+ways. First, the term being passed to `krun` can actually be symbolic. This
+is less frequently used because it requires the user to construct an AST
+that contains variables, something which our current parsing capabilities are
+not well-equipped to do. The second, more common, way of introducing symbolic
+terms into a configuration consists of writing rules where there exists an
+existentially qualified variable on the right-hand side of the rule that does
+not exist on the left-hand side of the rule.
+
+In order to prevent users from writing such rules by accident, K requires
+that such variables begin with the `?` prefix. For example, here is a rule
+that rewrites a constructor `foo` to a symbolic integer:
+
+```
+rule foo => ?X:Int
+```
+
+When this rule applies, a variable is introduced to the configuration, which
+then is unified against the rules that might apply in order to symbolically
+execute that configuration.
```

Comment:
`````suggestion
```
rule <k> foo => ?X:Int ... </k>
```

When this rule applies, a fresh variable is introduced to the configuration, which
then is unified against the rules that might apply in order to symbolically
execute that configuration.
`````

We should wrap this in `<k>` tags, because it's not logically sound to introduce fresh variables in function rules (which this looks like).